### ISSUE TITLE
Randomly miss the deviceType from MQTT

### ISSUE THREAD

**Issue by tagliala** [thingsboard/thingsboard#1471]
Hi,

I've noticed that sometimes the device type is not set by ThingsBoard. This happens randomly.

I'm using:
- ThingsBoard 2.3.0
- ThingsBoard Gateway 2.2.1rc
- Cassandra 3.11.3
- Mosquitto 1.5.5

If you think that this issue belongs to TB Gateway, please let me know I will post it in the proper repository

MQTT Config (timeseries omitted for brevity):
```json
{
  "brokers": [
    {
      "host": "mosquitto",
      "port": 1883,
      "ssl": false,
      "retryInterval": 3000,
      "credentials": {
        "type": "anonymous"
      },
      "mapping": [
        {
          "topicFilter": "newRE",
          "converter": {
            "type": "json",
            "filterExpression": "",
            "deviceNameJsonExpression": "newRE",
            "deviceTypeJsonExpression": "tracking",
            "timeseries": [
            ]
          }
        },
        {
          "topicFilter": "newOE",
          "converter": {
            "type": "json",
            "filterExpression": "",
            "deviceNameJsonExpression": "newOE",
            "deviceTypeJsonExpression": "tracking",
            "timeseries": [
            ]
          }
        },
        {
          "topicFilter": "newC",
          "converter": {
            "type": "json",
            "filterExpression": "",
            "deviceNameJsonExpression": "newC",
            "deviceTypeJsonExpression": "tracking",
            "timeseries": [
            ]
          }
        },
        {
          "topicFilter": "newH",
          "converter": {
            "type": "json",
            "filterExpression": "",
            "deviceNameJsonExpression": "newH",
            "deviceTypeJsonExpression": "tracking",
            "timeseries": [
            ]
          }
        },
        {
          "topicFilter": "newD",
          "converter": {
            "type": "json",
            "filterExpression": "",
            "deviceNameJsonExpression": "newD",
            "deviceTypeJsonExpression": "tracking",
            "timeseries": [
            ]
          }
        },
        {
          "topicFilter": "newM",
          "converter": {
            "type": "json",
            "filterExpression": "",
            "deviceNameJsonExpression": "newM",
            "deviceTypeJsonExpression": "tracking",
            "timeseries": [
            ]
          }
        },
        {
          "topicFilter": "newO",
          "converter": {
            "type": "json",
            "filterExpression": "",
            "deviceNameJsonExpression": "newO",
            "deviceTypeJsonExpression": "tracking",
            "timeseries": [
            ]
          }
        }
      ],
      "connectRequests": [
      ],
      "disconnectRequests": [
      ],
      "attributeRequests": [
      ],
      "attributeUpdates": [
      ],
      "serverSideRpc": [
      ]
    }
  ]
}
```

Log from ThingsBoard gateway:
```
16:57:02.940 [pool-4-thread-1] INFO  o.t.g.s.gateway.MqttGatewayService - [newC][tracking][3] Device connect event is reported to Thingsboard!
16:57:03.623 [pool-4-thread-1] INFO  o.t.g.s.gateway.MqttGatewayService - [newD][tracking][5] Device connect event is reported to Thingsboard!
16:57:04.145 [pool-4-thread-1] INFO  o.t.g.s.gateway.MqttGatewayService - [newH][tracking][8] Device connect event is reported to Thingsboard!
16:57:04.373 [pool-4-thread-1] INFO  o.t.g.s.gateway.MqttGatewayService - [newM][tracking][11] Device connect event is reported to Thingsboard!
16:57:04.532 [pool-4-thread-1] INFO  o.t.g.s.gateway.MqttGatewayService - [newO][tracking][15] Device connect event is reported to Thingsboard!
16:57:04.976 [pool-4-thread-1] INFO  o.t.g.s.gateway.MqttGatewayService - [newOE][tracking][20] Device connect event is reported to Thingsboard!
16:57:05.275 [pool-4-thread-1] INFO  o.t.g.s.gateway.MqttGatewayService - [newRE][tracking][24] Device connect event is reported to Thingsboard!
```

Screenshot from ThingsBoard
![image](https://user-images.githubusercontent.com/556268/52519111-e1e1ec00-2c56-11e9-82b0-a73902bea28d.png)

`newH` device should have `tracking` type.

**Comment by mfalkvidd** [comment]
On my Thingsboard instance, DeviceType is only updated when the connect event happens. The connect event only happens if there has ben a disconnect event, which seems to happen randomly (sometimes after about 1 minute, sometimes after 10+ minutes) after connect (when I test with mosquitto_pub which disconnects from the broker immediately after publishing the data).

Maybe that's what happens to you as well?

```
2019-04-10 15:51:42,093 [MQTT Call: 8e1f5b48-dd13-46bd-b55e-dd84c37f7f47] INFO  o.t.g.s.gateway.MqttGatewayService - [mftest3] Device Connected!
2019-04-10 15:51:42,789 [pool-2-thread-1] DEBUG o.t.g.service.MqttMessageSender - Sending message [{deviceId='mftest3', payload={"device":"mftest3","type":"mftest/v1"}, timestamp=0, topic='v1/gateway/connect'id=200e88f1-4e0b-4810-8
a4b-71622f43352b, messageId=120}]
```

Setting `<logger name="org.thingsboard" level="DEBUG" />` in` /usr/share/tb-gateway/conf/logback.xml` (and restarting tb-gateway before looking in `/var/log/tb-gateway/tb-gateway.log`) helped a lot when troubleshooting this.

**Comment by darkfader** [comment]
Can the device type be changed/checked on the server-side using the rule chain? I don't want some hacker to create unlimited devices with a gateway, or connect devices without proper permissions.
I am wondering if we can do any limiting on these connects... (security-wise, on the server side). Is it done by relations, permissions.. ?

I found out that an existing connection with the same client ID will be disconnected, so this makes testing more difficult. The solution is to use an intermediate MQTT broker there is only one connection. Also the device name / unique ID seems to be cached per connection.

**Comment by tagliala** [comment]
@mfalkvidd thanks for your help and sorry for this very late reply

> Maybe that's what happens to you as well?

Yes, it's the same issue as mine

```
2019-05-24 16:44:25,364 [pool-3-thread-1] INFO  o.t.g.s.gateway.MqttGatewayService - [Device_101][machine][7] Device connect event is reported to Thingsboard!
2019-05-24 16:44:25,418 [pool-3-thread-1] DEBUG o.t.g.s.gateway.MqttGatewayService - [Device_101][8] Device attributes were delivered!
2019-05-24 16:44:25,471 [pool-3-thread-1] INFO  o.t.g.s.gateway.MqttGatewayService - [Device_201][machine][13] Device connect event is reported to Thingsboard!
2019-05-24 16:44:25,493 [pool-3-thread-1] DEBUG o.t.g.s.gateway.MqttGatewayService - [15][Device_201] Device telemetry published to ThingsBoard!
2019-05-24 16:44:25,510 [pool-3-thread-1] DEBUG o.t.g.s.gateway.MqttGatewayService - [Device_201][14] Device attributes were delivered!
2019-05-24 16:44:25,608 [pool-3-thread-1] DEBUG o.t.g.s.gateway.MqttGatewayService - [12][Device_102] Device telemetry published to ThingsBoard!
2019-05-24 16:44:25,638 [pool-3-thread-1] INFO  o.t.g.s.gateway.MqttGatewayService - [Device_102][machine][10] Device connect event is reported to Thingsboard!
2019-05-24 16:44:25,644 [pool-3-thread-1] DEBUG o.t.g.s.gateway.MqttGatewayService - [9][Device_101] Device telemetry published to ThingsBoard!
2019-05-24 16:44:25,659 [pool-3-thread-1] DEBUG o.t.g.s.gateway.MqttGatewayService - [Device_102][11] Device attributes were delivered!
```

Device 102 didn't get the device type because the telemetry occurred before the connect event.

The problem is that those devices are dynamically created. In my above examples, I'm sending three times the very same message with different device id. 

Is there a way to prevent this and sent the connect before publishing attributes and telemetry?

**Comment by tmontanaro** [comment]
Hello, did anyone find a solution to this issue?

**Comment by tmontanaro** [comment]
Is it possible to overcome the problem by creating a Rule to change the Entity of device depending on an attribute in the msg? I tried to do it but no action allows to change the name or the type of a device. Is it right?

**Comment by ashvayka** [comment]
Hi @tmontanaro and everyone on this thread. Device Type is populated only once, when device is created. What is probably happening is that gateway receives a data sample with deviceName but with missing deviceType. Obviously it uses default. I don't see any other way how this can happen. Please note that Gateway "simulates" connect from device when it first receives data. So, you may receive a garbage data from some topic and automatically create device without device type. Maybe this is not your case, but one needs to be careful with list of subscribed topics.

**Comment by tmontanaro** [comment]
I did some further checks, and I noticed that if I force the name of the device to be lower than 10 characters the type is never missed.
Is it possible?
Since now I was using UUID4 as device names, but maybe they are too long?

**Comment by mykhailokornieiev** [comment]
As the community has not provided assistance on this matter over an extended period, I am closing this ticket due to inactivity. If it is still relevant, feel free to open a new ticket with a full description of your request.
